### DISCUSSION TITLE
Azure Devops 2 stages deploy - parametry?

### DISCUSSION THREAD

**Discussion by Mski89** [AzureCommunityPL/group#176] [Pomoc] [answered]
Cześć, 
zaczynam powoli wgryzać się w devopsa i natknąłem się na problem którego nie potrafię przeskoczyć.
W momencie gdy posiadam dwa srodowiska - prod, test - ktore posiadaja inne adresy do bazy, jak to przekazywac w devopsie? 
Sa Library Groups ktore moglbym przypiac do danego stage, ale nie wiem do czego moglo by sie to odnosic.

**Comment by kamil-mrzyglod** [comment]
Trochę mało informacji. Masz na myśli użycie stage w Releases? Z jakiej platformy korzystasz? 

Generalnie na poziomie Variables w Releases możesz dodać parametry o takiej samej nazwie a różnej wartości i ustawienie ich per stage.

**Reply by Mski89** [reply]
.Net Web App z baza w jednej RG (PROD)
.Net Web App z baza w drugiej RG (TEST)
connection string do bazy jest inny dla kazdego srodowiska (user, paasword).
Tworzac release pipeline jak moge wskazac ze "teraz deploy idzie na proda wiec masz skorzystac z connection stringa do proda" 

Dodam ze nie robie yamlem tylko klasycznym edytorem.

**Reply by kamil-mrzyglod** [reply]
Robisz stage TEST i stage PROD:

![image](https://user-images.githubusercontent.com/11725731/101896637-92014280-3ba9-11eb-8af4-ae013ad2c7a1.png)

W Variables definiujesz zmienną per stage:

![image](https://user-images.githubusercontent.com/11725731/101896734-b3fac500-3ba9-11eb-9800-7ebff5ab9a3c.png)

Per stage zmienna $(foo) będzie miała wartość zależną o faktycznej fazy deploymentu.

**Reply by Mski89** [reply]
Jeśli dobrze zoruzmialem tu wpisuje connection string do odpowiedniej bazy i przypisuje per stage tak ?

**Reply by wi5nia** [reply]
Connection string w Azure DevOps? No nie... to powinno siedziec w KeyVault

**Reply by kamil-mrzyglod** [reply]
@wi5nia na samym końcu powinno - rozwinięcie tematu podrzucił @mgrabarz z Variable Groups.

**Comment by mgrabarz** [comment, marked as answer]
Jak dla mnie jest to klasyczn use-case dla Variable Groups. 
- Definiujesz dwie z tymi samymi kluczami, różnymi wartościami.
- W stage DEV podpinasz pierwszą.
- W stage PROD podpinasz drugą.

**Reply by Mski89** [reply]
Dzieki @mgrabarz. 
Rozumiem ze to tyle po stronie devopsa i trzeba dopisac jakis IF w kodzie aplikacji ?

**Reply by kamil-mrzyglod** [reply]
@Mski89 nic po stronie appki nie dopisujesz - https://docs.microsoft.com/en-us/azure/devops/pipelines/library/variable-groups?view=azure-devops&tabs=yaml

**Reply by wi5nia** [reply]
Ale nie trzymajcie connection stringow w DevOps... no c'mon

**Reply by Mski89** [reply]
@wi5nia  spokojnie. Wiem od czego jest Keyvault ale pytanie dotyczyło rozróżnienia środowisk. Bo nawet keyvault bedzie 1 dla prod a drugi dla test :) Musze załapac jak odróżnić srodowiska :)

**Reply by wi5nia** [reply]
Ale jak masz KV no to chyba jak robisz deploy to wiesz na jakie srodowisko i ono ma swojego KV i z niego bedzie korzystalo :) ale spoko... kumam...

**Reply by Mski89** [reply]
"In the Variable groups page, enable Link secrets from an Azure key vault as variables." 
Fakt. Szybciej szło to załapac używająć KV jako przykładu :P 
Dzięki @wi5nia @kamil-mrzyglod @mgrabarz !!
